File: python/spiral-matrix-iv.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

DIRECTIONS = [(0, 1), (1, 0), (0, -1), (-1, 0)]

class Solution:
    def init_spiral_matrix(m: int, n: int) -> List[List[int]]:
        spiral_matrix = []
        for i in range(m):
            spiral_matrix.append([-1] * n)
        return spiral_matrix

    def get_direction_index(i: int, j: int, m: int, n: int, direction_index: int, spiral_matrix: List[List[int]]) -> int:
        direction = DIRECTIONS[direction_index]
        i += direction[0]
        j += direction[1]
        if i == m or j == n or spiral_matrix[i][j] != -1:
            direction_index = (direction_index + 1) % 4
        return direction_index

    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        spiral_matrix = Solution.init_spiral_matrix(m, n)
        i, j = 0, 0
        direction_index = 0
        direction = DIRECTIONS[direction_index]
        while head:
            spiral_matrix[i][j] = head.val
            direction_index = Solution.get_direction_index(i, j, m, n, direction_index, spiral_matrix)
            direction = DIRECTIONS[direction_index]
            i += direction[0]
            j += direction[1]
            head = head.next
        return spiral_matrix

File: go/spiral-matrix-iv.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

 func InitMatrix(m int, n int) [][]int {
    spiralMatrixOutput := make([][]int, m)
    for i := range spiralMatrixOutput {
        spiralMatrixOutput[i] = make([]int, n)
        for j := range spiralMatrixOutput[i] {
            spiralMatrixOutput[i][j] = -1
        }
    }
    return spiralMatrixOutput
}

func IsInRange(i int, m int) bool {
    return 0 <= i && i < m
}

func GetDirectionIndex(i int, j int, m int, n int, directionIndex int, spiralMatrixOutput [][]int, directions [][]int) int {
    direction := directions[directionIndex]
    i += direction[0]
    j += direction[1]
    if !IsInRange(i, m) || !IsInRange(j, n) || spiralMatrixOutput[i][j] != -1 {
        directionIndex = (directionIndex + 1) % 4
    }
    return directionIndex
}

func spiralMatrix(m int, n int, head *ListNode) [][]int {
    directions := [][]int{
        {0, 1},
        {1, 0},
        {0, -1},
        {-1, 0},
    }
    spiralMatrixOutput := InitMatrix(m, n)
    i, j, directionIndex := 0, 0, 0
    for head != nil {
        spiralMatrixOutput[i][j] = head.Val
        directionIndex = GetDirectionIndex(i, j, m, n, directionIndex, spiralMatrixOutput, directions)
        direction := directions[directionIndex]
        i += direction[0]
        j += direction[1]
        head = head.Next
    }
    return spiralMatrixOutput
}
